pkg/ingester/ingester_v2.go

```diff
@@ -695,37 +692,24 @@ func (i *Ingester) v2MetricsForLabelMatchers(ctx context.Context, req *client.Me
 		}
 
 		seriesSet := q.Select(false, nil, matchers...)
```

Comment:
Sets with series are not sorted, so merging them later doesn't work properly. This can be shown by using modified request in the benchmark/test:

```
	req := &client.MetricsForLabelMatchersRequest{
		StartTimestampMs: now,
		EndTimestampMs:   now,
		MatchersSet: []*client.LabelMatchers{
			{Matchers: []*client.LabelMatcher{
				{Type: client.REGEX_MATCH, Name: model.MetricNameLabel, Value: "test.*"},
			}},
			{Matchers: []*client.LabelMatcher{
				{Type: client.REGEX_MATCH, Name: model.MetricNameLabel, Value: "test.*0"}, // ending with 0
			}},
		},
	}
```

This should still return all `numSeries` series, but returns more of them.

Can you rerun benchmarks with sorting, and see whether it is still an improvement or not?

Reply:
Wow, thanks for spotting it! I copied the logic from Prometheus:
https://github.com/prometheus/prometheus/blob/b521612042ec87103088b45b7b5dfee3bb8dc732/web/api/v1/api.go#L604-L608

Either Prometheus /series API endpoint doesn't guarantee any deduplication, or there's a bug in Prometheus too. I will open an issue there.

Back to us, I modified the logic to use `.Select(true, ...)` (sorted) for now and re-run the benchmark (updated PR description). There's still some benefit, less then before.

Reply:
Opened this issue in Prometheus: https://github.com/prometheus/prometheus/issues/7801. Will follow up this PR accordingly, if any change is required.